stablehlo/reference/ProcessGrid.cpp

```diff
@@ -162,6 +197,21 @@ ProcessGroups ProcessGrid::flattenedIds(
 
 void ProcessGrid::outfeed(ArrayRef<Tensor> inputs) { outfeed_.push(inputs); }
 
+SmallVector<Tensor> ProcessGrid::recv(ChannelId channelId,
+                                      ProcessId processId) {
+  sendRecvReady_.insert(channelId);
+  sendRecvConditions_[channelId].notify_one();
```

Comment:
Is there an assumption that `send` will be waiting by the time this code is reached? Wondering if this is a safe assumption to make. I.e. is it possible that this notify_one either notifies nothing, or notifies the recv thread itself, both of which would cause deadlock/timeout

Reply:
Can test this by adding a thread sleep for a second to the send impl.

One possible way to avoid this could be a separate `sendCondition` and `recvCondition`. This doesn't resolve the "notifies nothing" potential issue, but does avoid a self-wakeup if that turns out to be a problem

Reply:
There are two cases:

1. recv arrives first:
* Insert `channelId` to `sendRecvReady_`.
* `notify_one` notifies nothing and locks `sendRecvChannels_[channelId].mutex` or notifies send to continue.
  * If `notify_one` notifies nothing and recv will wait. Send will then gain control of lock and see that `sendRecvReady_.find(channelId) != sendRecvReady_.end()` condition is satisfied and move on (releasing the lock), and `sendRecvChannels_[channelId].result` is populated and notify, at which point recv condition is also satisfied.
  * If `notify_one` notifies send, `sendRecvChannels_[channelId].result` is populated and notify, at which point recv condition is also satisfied.
2. send arrives first:
* Locks `sendRecvChannels_[channelId].mutex` and wait (release the lock). Recv inserts to `sendRecvReady_` and `notify_one` which wakes send and the condition is satisfied.

I think the "notify nothing" part is okay to have, and it won't cause deadlocks.

Reply:
As long as `wait_for` doesn't block if the precondition is true when the statement is reached, this probably isn't an issue, which seems to be the case

Reply:
Let's still move the lock in recv to the top of the method. That way it is guarantees that either send gets the lock first, and joins the waiting queue, or recv gets the lock first and send's wair_for confition is guaranteed to be true by the time execution reaches there

Reply:
According to the [cpp reference](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one) for `notify_one`,

> The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s); in fact doing so is a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock.

Some implementations will recognize this and transfer the ownership, but otherwise notifying the other thread while having control of the lock would immediately leave the other thread waiting again.

Here is a [diagram](https://www.tensorflow.org/xla/operation_semantics#send) that may complement my explanation, but essentially, since send is immediately trying to acquire the lock, if it has the lock, it will proceed to wait (and release lock). If it doesn't have the lock, it will wait until the lock is free to acquire. This allows recv to freely insert `channelId` to `sendRecvReady_` at any time, and since send is already already waiting on the condition_variable.

Before I go deeper into the explanation (at this point this little code snippet seems to cause more confusion!), I'm questioning whether the `notify_one` is actually needed. If we remove `notify_one` from both functions, it might be easier to reason about the parallel execution and I think it's unnecessary. Before I make this comment even longer, can you look at the updated code and let's discuss :)

Reply:
But I agree that it doesn't hurt to be extra cautious and makes reading the two implementation easier to reason about:

There are now two cases:
1. Recv acquires lock first.
2. Send acquires lock first.

Reply:
LGTM! I had seen that clause before about "does not need to hold the lock on the same mutex as the one held by the waiting thread(s)", but I think this is a fair means of doing it. Our other ops notify the same way so this is consistent. If we end up finding this to be an issue on some hardware/compiler (hasn't happened yet), then we can re-think a strategy and fix them all together